### PR TITLE
feat: api migration

### PR DESCRIPTION
**Pull request by volodymyr-chekyrta**
Migration API endpoints from the RG API plugin to the edx-platform upstream.

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @volodymyr-chekyrta! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

**Comment by omerhabib26** [comment]
I have started its review

File: course/src/main/java/org/openedx/course/data/repository/CourseRepository.kt [hunk 1]
```diff
@@ -51,7 +51,7 @@ class CourseRepository(
     suspend fun preloadCourseStructure(courseId: String) {
         val response = api.getCourseStructure(
             "stale-if-error=0",
-            "v1",
+            "v3",
```
Comment (omerhabib26) [inline course/src/main/java/org/openedx/course/data/repository/CourseRepository.kt:54]: we can define it config.yaml

File: app/src/main/java/org/openedx/app/AppRouter.kt [hunk 2]
```diff
@@ -113,13 +112,9 @@ class AppRouter : AuthRouter, DiscoveryRouter, DashboardRouter, CourseRouter, Di
...
         fm: FragmentManager,
         title: String,
-        coursewareAccess: CoursewareAccess,
         auditAccessExpires: Date?
     ) {
-        replaceFragment(
-            fm,
-            NoAccessCourseContainerFragment.newInstance(title, coursewareAccess, auditAccessExpires)
-        )
+        replaceFragment(fm, NoAccessCourseContainerFragment.newInstance(title, auditAccessExpires))
```
Comment (omerhabib26) [inline app/src/main/java/org/openedx/app/AppRouter.kt:117]: IMO, we can remove the `auditAccessExpires` parameter as well

**Review by omerhabib26 — CHANGES_REQUESTED** [review]
CourseDetailInfo isn't working else only some minor nits.
Suggestion: IMO, For the APIs we can define their versions in `Config.yaml`.

File: core/src/main/java/org/openedx/core/data/api/CourseApi.kt [hunk 3]
```diff
@@ -28,15 +29,15 @@ interface CourseApi {
         )
     ): CourseList
 
-    @GET("/mobile_api_extensions/v1/courses/{course_id}")
+    @GET("/api/mobile/v3/course_info/{course_id}/info")
```
[the remark this answers is not in the document]
Comment (volodymyr-chekyrta) [reply core/src/main/java/org/openedx/core/data/api/CourseApi.kt:32]: It's working well on my end
image

Can you make sure you've got the latest changes from upstream?
https://github.com/openedx/edx-platform/pull/33646

File: app/src/main/java/org/openedx/app/AppRouter.kt (hunk 2 above)
Comment (volodymyr-chekyrta) [reply app/src/main/java/org/openedx/app/AppRouter.kt:117]: It looks like the `NoAccessCourseContainerFragment` uses this parameter.

image

File: core/src/main/java/org/openedx/core/data/model/CourseEnrollments.kt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package org.openedx.core.data.model
+
+import com.google.gson.annotations.SerializedName
+
+data class CourseEnrollments(
+    @SerializedName("enrollments")
+    val enrollments: DashboardCourseList
```
[the remark this answers is not in the document]
Comment (volodymyr-chekyrta) [reply core/src/main/java/org/openedx/core/data/model/CourseEnrollments.kt:7]: Right now, applications aren't using this class.
I'd like to avoid wasting application resources parsing an unused class.

File: core/src/main/java/org/openedx/core/data/model/CourseEnrollments.kt [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+)
```
[the remark this answers is not in the document]
Comment (volodymyr-chekyrta) [reply core/src/main/java/org/openedx/core/data/model/CourseEnrollments.kt]: ✅

File: core/src/main/java/org/openedx/core/data/api/CourseApi.kt [hunk 6]
```diff
@@ -6,19 +6,20 @@ import retrofit2.http.*
 
 interface CourseApi {
 
-    @GET("/mobile_api_extensions/v1/users/{username}/course_enrollments")
+    @GET("/api/mobile/v3/users/{username}/course_enrollments/")
```
[the remark this answers is not in the document]
Comment (volodymyr-chekyrta) [reply core/src/main/java/org/openedx/core/data/api/CourseApi.kt:9]: I believe it's a great suggestion 👍👍👍
It might be beneficial to handle this improvement in a separate pull request to include a user API version to the config as well.
What do you think?

File: core/src/main/java/org/openedx/core/data/api/CourseApi.kt [hunk 7]
```diff
@@ -28,15 +29,15 @@ interface CourseApi {
...
+    @GET("/api/mobile/v3/course_info/{course_id}/info")
     suspend fun getCourseDetail(
         @Path("course_id") courseId: String?,
         @Query("username") username: String? = null,
         @Query("is_enrolled") isEnrolled: Boolean = true,
     ): CourseDetails
 
     @GET(
-        "/mobile_api_extensions/{api_version}/blocks/?" +
+        "/api/mobile/{api_version}/course_info/blocks/?" +
```
[the remark this answers is not in the document]
Comment (volodymyr-chekyrta) [reply core/src/main/java/org/openedx/core/data/api/CourseApi.kt:40]: Same, improvements with versions in a separate PR?

File: course/src/main/java/org/openedx/course/data/repository/CourseRepository.kt (hunk 1 above)
Comment (volodymyr-chekyrta) [reply course/src/main/java/org/openedx/course/data/repository/CourseRepository.kt:54]: Same, improvements with versions in a separate PR?

**Comment by volodymyr-chekyrta** [comment]
@omerhabib26, thank you for review!
Ready for another pass.

File: core/src/main/java/org/openedx/core/data/api/CourseApi.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (omerhabib26) [reply core/src/main/java/org/openedx/core/data/api/CourseApi.kt:9]: sure,
@moiz994 can you please create a ticket for these improvements ?

File: app/src/main/java/org/openedx/app/AppRouter.kt (hunk 2 above)
Comment (omerhabib26) [reply app/src/main/java/org/openedx/app/AppRouter.kt:117]: [Router.NavigateToNoAccess](https://github.com/openedx/openedx-app-android/pull/171/files/b26610295e36760ffb1b4cc81ac2c22519283cdc#diff-afd03945c123e5bcb0cdc134a628d8ea040dbb3bf0a7567c133261d6d3bf4151R112) is always getting null for `auditAccessExpires`

**Review by omerhabib26 — APPROVED** [review]
Only a minor suggestion else LGTM 🚀

**Review by omerhabib26 — CHANGES_REQUESTED** [review]
`/api/mobile/v3/course_info/` API not working and giving 400

File: app/src/main/java/org/openedx/app/AppRouter.kt (hunk 2 above)
Comment (volodymyr-chekyrta) [reply app/src/main/java/org/openedx/app/AppRouter.kt:117]: Got it, I'll check it

Comment (volodymyr-chekyrta) [reply app/src/main/java/org/openedx/app/AppRouter.kt]: Yeah, you're right. I found the `null` value and removed the `auditAccessExpires` argument

**Comment by volodymyr-chekyrta** [comment]
> `/api/mobile/v3/course_info/` API not working and giving 400

The API has been changed and PR is now ready for review

**Review by k1rill — APPROVED** [review]

File: course/src/main/java/org/openedx/course/presentation/container/CourseContainerFragment.kt [hunk 8]
```diff
@@ -66,18 +66,14 @@ class CourseContainerFragment : Fragment(R.layout.fragment_course_container) {
...
-                } else {
-                    router.navigateToNoAccess(
-                        requireActivity().supportFragmentManager,
-                        courseTitle,
-                        coursewareAccess,
-                        null
-                    )
-                }
+        viewModel.dataReady.observe(viewLifecycleOwner) { isReady ->
+            if (isReady == true) {
```
Comment (omerhabib26) [inline course/src/main/java/org/openedx/course/presentation/container/CourseContainerFragment.kt:70]: This condition doesn't work properly with the expired course and shows an API loading error.
CourseStatusInfo, CourseDates, and Topis API are failing due to course expiry. I guess we can update this change with the following for now. We also need to update the message on NoAccessScreen accordingly. Right now it showing `The Course hasn't started yet` for every case.
```
        viewModel.dataReady.observe(viewLifecycleOwner) { coursewareAccess ->
            if (coursewareAccess != null && coursewareAccess.hasAccess) {
                initViewPager()
            } else {
                router.navigateToNoAccess(
                    requireActivity().supportFragmentManager,
                    courseTitle
                )
            }
        }

File: course/src/main/java/org/openedx/course/presentation/container/NoAccessCourseContainerFragment.kt [hunk 9]
```diff
@@ -36,19 +35,6 @@ import org.openedx.course.R as courseR
 
 class NoAccessCourseContainerFragment : Fragment() {
 
-    private var courseTitle = ""
```
Comment (omerhabib26) [inline course/src/main/java/org/openedx/course/presentation/container/NoAccessCourseContainerFragment.kt:39]: I guess we can revert this change and handle it properly

**Review by omerhabib26 — CHANGES_REQUESTED** [review]
Some APIs are failing in case of course expiry we need to handle them properly

**Comment by volodymyr-chekyrta** [comment]
> Some APIs are failing in case of course expiry we need to handle them properly

I know about this case. We encountered a similar problem in the iOS PR https://github.com/openedx/openedx-app-ios/pull/213 and decided to address it in a separate PR. 
At the moment, I'm uncertain how to handle this case since we don't have `coursewareAccess` data.
It's possible that addressing this issue will require modifying the API. 
Therefore, I suggest we tackle this later in a separate PR to avoid getting blocked now while waiting for the API.

**Review by omerhabib26 — APPROVED** [review]

File: core/src/main/java/org/openedx/core/data/api/CourseApi.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (omerhabib26) [reply core/src/main/java/org/openedx/core/data/api/CourseApi.kt:9]: No need to create the ticket for now, we will handle it later

**Comment by openedx-webhooks** [comment]
@volodymyr-chekyrta 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/openedx-app-android&entry.867055334=https://github.com/openedx/openedx-app-android/pull/171&entry.1484655318=https://github.com/volodymyr-chekyrta&entry.752974735=2023-12-19+15:41&entry.1917517419=2024-01-05+09:43&entry.2133058324=Yes) so we can improve your experience in the future.
